### PR TITLE
New options page to add domain regexes to a whitelist

### PR DESCRIPTION
**Pull request by li-xin-yi**
I really appreciate this project and look forward to a new version that I can use on more websites more flexibly. I attempted to modify a few lines of code as below:

1. Refactor the options page by React, and support users to edit their own domain lists on which Codeium is allowed to complete code. The default list is the following 11 regular expressions, extracted from `script.ts`

```
https:\/\/colab.research\.google\.com\/.*
https:\/\/(.*\.)?stackblitz\.com\/.*
https:\/\/(.*\.)?deepnote\.com\/.*
https:\/\/(.*\.)?(databricks\.com|azuredatabricks\.net)\/.*
https:\/\/(.*\.)?quadratichq\.com\/.*
https?:\/\/(.*\.)?jsfiddle\.net(\/.*)?
https:\/\/(.*\.)?codepen\.io(\/.*)?
https:\/\/(.*\.)?codeshare\.io(\/.*)?
https:\/\/console\.paperspace\.com\/.*\/notebook\/.*
https?:\/\/www\.codewars\.com(\/.*)?
https:\/\/(.*\.)?github\.com(\/.*)?
```

image

2. Test the feature on some web pages with the online editor, e.g., leetcode.com, codepod.io, and some web apps I deployed on local env with URL starting by localhost. It works when I write a correct regex for them. And also in reverse, removing some default regexes above can disable the code completion of Codeium on those websites.
3. Though it extends the application scenarios for Codeium further, it may not work on many websites even if they apply Monaco or CodeMirror as their online editors (for example, online vs code/github.dev, codesandbox). The rules are too rough to support them well for now.

### CODE REVIEW

**Comment by pqn** [comment]
> Though it extends the application scenarios for Codeium further, it may not work on many websites even if they apply Monaco or CodeMirror as their online editors (for example, online vs code/github.dev, codesandbox). The rules are too rough to support them well for now.

We have an internal issue tracking github.dev -- this would require a modification of our vscode extension to support the web environment.

I've previously looked at CodeSandbox. I think it doesn't work for a similar reason as described in https://github.com/Exafunction/codeium-chrome/issues/4#issuecomment-1519231767

File: src/popup.ts [hunk 1]
```diff
@@ -6,7 +6,7 @@ import { getStorageItem, setStorageItem } from './storage';
 document.getElementById('login')?.addEventListener('click', openAuthTab);
 
 document.getElementById('go-to-options')?.addEventListener('click', () => {
-  chrome.runtime.openOptionsPage();
+  chrome.tabs.create({ url: 'chrome://extensions/?options=' + chrome.runtime.id });
```
Comment (pqn) [inline src/popup.ts:9]: What's the reason for this change?

File: src/script.ts [hunk 2]
```diff
@@ -150,27 +172,47 @@ const SUPPORTED_CODEMIRROR_NONGLOBAL_SITES = [
...
+  }, 100);
+};
+
+getWhiteList(extensionId).then((waitList) => {
+  for (const addr of waitList ?? []) {
+    const host = new RegExp(addr);
+    if (host.test(window.location.href)) {
+      // the url matches the whitelist
+      addMonacoInject();
+      addCodeMirrorInject();
```
Comment (pqn) [inline src/script.ts:213]: Is it possible that both `addCodeMirrorInject` and `addCodeMirror5Inject` will run on the same site?

File: src/component/Options.tsx [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+        value={text}
+        multiline
+        rows={7}
+        onChange={(e) => setText(e.target.value)}
+      ></TextField>
+
+      <Button
+        variant="text"
+        onClick={() => {
+          setStorageItem('whitelist', text.split('\n'));
```
Comment (pqn) [inline src/component/Options.tsx:53]: Let's validate that `new RegExp(...)` doesn't throw before saving.

File: src/component/Options.tsx [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+            verticalAlign: 'bottom',
+          }}
+          fontSize="large"
+        />
+      </Typography>
+      <Typography variant="body2">
+        Domains to allow auto-completion, double-click to edit, use regex format
+      </Typography>
+
+      <TextField
```
Comment (pqn) [inline src/component/Options.tsx:40]: Might look better as a mono font.

**Review by pqn — COMMENTED** [review]
Really appreciate the PR! I just left a few comments.

File: src/popup.ts (hunk 1 above)
Comment (li-xin-yi) [reply src/popup.ts:9]: I don't think the options page is large enough to take up one whole tab, so use this link to go to the extension detail page and pop up the options window.

File: src/script.ts (hunk 2 above)
Comment (li-xin-yi) [reply src/script.ts:213]: Yep, I tried on Code Share and it gave completion twice. Thanks for pointing the bug out.

**Comment by li-xin-yi** [comment]
@pqn 

Thank you so much for reviewing this PR and valuable suggestions. I pushed some new commits to complete:

1. Change the font of domain inputs to "monospace"
2. Validate the domain inputs, add toast notification on success/error when clicking on the save button
3. Fix the duplicated injection issue

Can you take a look and give more feedback?

File: src/component/Options.tsx [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+const EditableList = () => {
+  const [text, setText] = useState('');
+  const [open, setOpen] = useState(false);
+  const [severity, setSeverity] = useState<'success' | 'error'>('success');
+  const [message, setMessage] = useState('');
+
+  useEffect(() => {
+    (async () => {
+      const whitelist = await getStorageItem('whitelist');
+      setText(whitelist.join('\n'));
```
Comment (pqn) [inline src/component/Options.tsx:24]: Prefer calling this allowlist instead of whitelist everywhere.

File: static/options.html [hunk 6]
```diff
@@ -1,30 +1,13 @@
...
-        <button id="tokenpage" class="button">Open token page</button>
-        <button id="entertoken" class="button">Enter token</button>
-      </div>
-    </div>
-
+    <div id="codeium-options"></div>
     <script src="options.js"></script>
   </body>
-</html>
+</html>
```
Comment (pqn) [inline static/options.html:13]: Merge/rebase on main and some additional CI checks will run. I think they might complain about missing `\n` here.

File: src/component/Options.tsx [hunk 7]
```diff
@@ -0,0 +1,183 @@
+import React, { createRef, useEffect, useRef, useState } from 'react';
```
Comment (pqn) [inline src/component/Options.tsx:1]: image

Visually this part seems a bit crowded. Maybe we can do without the icon and the redundant "Domain Regexes" text.

File: src/component/Options.tsx [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+          setSeverity('success');
+          setMessage('Saved successfully');
+          setOpen(true);
+        }}
+        sx={{
+          float: 'right',
+          textTransform: 'none',
+        }}
+      >
+        Save <SaveAltIcon />
```
Comment (pqn) [inline src/component/Options.tsx:83]: It's ambiguous from the UI if the Save button will also save the token field or not.

File: src/script.ts [hunk 9]
```diff
@@ -111,37 +128,43 @@ const SUPPORTED_CODEMIRROR_SITES = [
...
-      const jupyterState = jupyterInject(extensionId, this.Jupyter);
-      addListeners(cm as CodeMirror, jupyterState.codeMirrorManager);
-    } else {
-      let injectCodeMirror = false;
-      let multiplayer = false;
-      for (const pattern of SUPPORTED_CODEMIRROR_SITES) {
-        if (pattern.pattern.test(window.location.href)) {
+let injectCodeMirror = false;
+
+const addCodeMirrorInject = () =>
```
Comment (pqn) [inline src/script.ts:133]: Let's call this one `addCodeMirror5GlobalInject` and the other one `addCodeMirror5LocalInject`, since both only work on CodeMirror 5.

**Review by pqn — COMMENTED** [review]
Thanks, left some additional comments.

File: src/component/Options.tsx [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+        Whitelist{' '}
+        <ChecklistRtlIcon
+          sx={{
+            verticalAlign: 'bottom',
+          }}
+          fontSize="large"
+        />
+      </Typography>
+      <Typography variant="body2">
+        Domains to allow auto-completion, double-click to edit, use regex format
```
Comment (pqn) [inline src/component/Options.tsx:40]:
```suggestion
        Domains to allow auto-completion. Use one regex per line.
```

File: src/component/Options.tsx [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+            fontFamily: 'monospace',
+          },
+        }}
+        onChange={(e) => setText(e.target.value)}
+      ></TextField>
+
+      <Button
+        variant="text"
+        onClick={() => {
+          const lst = text.split('\n').map((x) => x.trim());
```
Comment (pqn) [inline src/component/Options.tsx:62]: Filter empty strings (so we can tolerate newlines as spacing).

**Review by pqn — COMMENTED** [review]
I forgot to mention earlier, but let's have a `Reset` button for the regex field to restore the default settings.

**Comment by li-xin-yi** [comment]
@pqn Thanks for your comment. I updated the code as you suggested. Now, the new UI looks like:

image

I'm uncertain whether it should be "allow list" or "allowed list" (since I'm not a native English speaker...)

File: src/component/Options.tsx [hunk 12]
```diff
@@ -0,0 +1,188 @@
...
+  useEffect(() => {
+    (async () => {
+      const allowList = (await getStorageItem('allowList')) ?? [];
+      setText(allowList.join('\n'));
+    })();
+  }, []);
+
+  return (
+    <>
+      <Typography variant="h6">Allow List</Typography>
```
Comment (pqn) [inline src/component/Options.tsx:30]:
```suggestion
      <Typography variant="h6">Allowlist</Typography>
```
(Similar changes elsewhere.)

File: src/component/Options.tsx [hunk 13]
```diff
@@ -0,0 +1,188 @@
...
+          setSeverity('success');
+          setMessage('Saved successfully');
+          setOpen(true);
+        }}
+        sx={{
+          float: 'right',
+          textTransform: 'none',
+        }}
+      >
+        Save the Allow List
```
Comment (pqn) [inline src/component/Options.tsx:77]:
```suggestion
        Save Allowlist
```

File: static/options.html (hunk 6 above)
Comment (pqn) [reply static/options.html]: Looks like we still need the merge/rebase (CI is still failing)

**Comment by pqn** [comment]
Looks like the rebase or merge had some issue and the commits are showing up in the PR history.

**Comment by li-xin-yi** [comment]
Update: add the reset button and remove the exit button.

**Review by pqn — APPROVED** [review]
One small change I missed, otherwise this looks good to merge!

File: src/monacoCompletionProvider.ts [hunk 14]
```diff
@@ -62,7 +62,7 @@ declare module 'monaco-editor' {
 }
 
 export const OMonacoSite = {
-  UNSPECIFIED: 0,
```
Comment (pqn) [inline src/monacoCompletionProvider.ts:65]: Sorry, to be more specific, let's keep UNSPECIFIED = 0 and add CUSTOM as a new value.

Comment (li-xin-yi) [reply src/monacoCompletionProvider.ts]: No problem, I recovered it.

**Comment by pqn** [comment]
Thanks! 🎉
